### PR TITLE
functional: add vfio test

### LINKED ISSUE #2451: Add functional VFIO tests

### PR DESCRIPTION
**Pull request by devimc**
Add functional VFIO test. The test will start a new VM with an extra ethernet
device that will be hotplugged (passthrough) in a kata container running in
the VM (nested virtualization).
Nested virtualization can be slow, hence a functional test was implemented
in order to avoid timeouts using container engines, for example docker.

fixes #2451

Signed-off-by: Julio Montes <julio.montes@intel.com>

### CODE REVIEW

File: .ci/jenkins_job_build.sh [hunk 1]
```diff
@@ -180,6 +180,11 @@ else
 	echo "WARN: Kata runtime is not installed"
 fi
 
+if [ -n "${VFIO_CI}" ]; then
+	"${ci_dir_name}/run.sh"
+	exit $?
```
Comment (jodh-intel) [inline .ci/jenkins_job_build.sh:185]: The `exit` won't get called strictly due to the `set -e`, so you could just simplify to:

```sh
[ -n "${VFIO_CI}" ] && "${ci_dir_name}/run.sh"
```

File: .ci/vfio_jenkins_job_build.sh [hunk 2]
```diff
@@ -0,0 +1,276 @@
...
+# that supports VFIO, then run VFIO functional tests
+
+set -x
+
+cidir=$(dirname "$0")
+
+source /etc/os-release || source /usr/lib/os-release
+source "${cidir}/lib.sh"
+
+[ "${NAME}" == "Ubuntu" ] || die "Unsupported distro ${NAME}. Please use Ubuntu"
```
Comment (jodh-intel) [inline .ci/vfio_jenkins_job_build.sh:18]: Could you add a comment / GitHub issue URL explaining why this is Ubuntu-specific?

File: .ci/vfio_jenkins_job_build.sh [hunk 3]
```diff
@@ -0,0 +1,276 @@
...
+vm_port="10022"
+data_dir="/tmp/kata-containers/vfio-test"
+ssh_key_file="${data_dir}/key"
+
+mkdir -p "${data_dir}"
+
+trap cleanup EXIT
+
+cleanup() {
+	sudo killall -9 qemu-system-x86_64
```
Comment (jodh-intel) [inline .ci/vfio_jenkins_job_build.sh:30]: Hardcoded architecture (here and below). Could you use `qemu-system-$(uname -i)` (or similar) instead?

File: .ci/vfio_jenkins_job_build.sh [hunk 4]
```diff
@@ -0,0 +1,276 @@
...
+	xorriso -as mkisofs -R -V config-2 -o "${iso_file}" "${iso_data_dir}"
+}
+
+pull_fedora_cloud_image() {
+	fedora_img="$1"
+	fedora_img_cache="${fedora_img}.cache"
+	fedora_version=30
+
+	if [ ! -f "${fedora_img_cache}" ]; then
+		curl -Lk "https://download.fedoraproject.org/pub/fedora/linux/releases/${fedora_version}/Cloud/x86_64/images/Fedora-Cloud-Base-${fedora_version}-1.2.x86_64.raw.xz" -o "${fedora_img_cache}.xz"
```
Comment (jodh-intel) [inline .ci/vfio_jenkins_job_build.sh:164]: More architecture-specific code.

**Comment by devimc** [comment]
@jodh-intel changes applied, thanks

@amshinde @chavafg @GabyCT ready for review, thanks

File: .ci/vfio_jenkins_job_build.sh [hunk 5]
```diff
@@ -0,0 +1,294 @@
...
+		if [ -n "${no_proxy}" ]; then
+			environment+="
+    export no_proxy=\"${no_proxy}\"
+    export NO_PROXY=\"${no_proxy}\"
+    export http_proxy=\"${http_proxy}\"
+    export HTTP_PROXY=\"${http_proxy}\"
+    export https_proxy=\"${https_proxy}\"
+    export HTTPS_PROXY=\"${https_proxy}\""
+		fi
+	fi
```
Comment (cmaf) [inline .ci/vfio_jenkins_job_build.sh:74]: @devimc is this formatting off or is like this because you are adding it to a file?

Comment (devimc) [reply .ci/vfio_jenkins_job_build.sh:74]: it's because I'm adding this to a yaml file, so the format is important

**Pull request merged**